File: samples/lifecycle_cpp/test/test_lifecycle_node.cpp
#include <gtest/gtest.h>
#include <memory>

#include "rclcpp/rclcpp.hpp"
#include "rde_lifecycle_cpp.hpp"

// Mock the main code for testing the RdeLifecycleNode
class RdeLifecycleNodeTest : public ::testing::Test {
protected:
  void SetUp() override {
    rclcpp::init(0, nullptr);
  }

  void TearDown() override {
    if (rclcpp::ok()) {
      rclcpp::shutdown();
    }
  }
};

// Test that lifecycle node can be instantiated
TEST_F(RdeLifecycleNodeTest, NodeInstantiation) {
  auto node = std::make_shared<RdeLifecycleNode>("test_node");
  EXPECT_NE(node, nullptr);
  EXPECT_EQ(node->get_name(), std::string("test_node"));
}

// Test that on_configure reBturns SUCCESS
TEST_F(RdeLifecycleNodeTest, OnConfigureReturnsSuccess) {
  auto node = std::make_shared<RdeLifecycleNode>("test_node");
  rclcpp_lifecycle::State state;
  auto result = node->on_configure(state);
  EXPECT_EQ(
    result,
    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS
  );
}

// Test that on_activate returns SUCCESS
TEST_F(RdeLifecycleNodeTest, OnActivateReturnsSuccess) {
  auto node = std::make_shared<RdeLifecycleNode>("test_node");
  rclcpp_lifecycle::State state;
  auto result = node->on_activate(state);
  EXPECT_EQ(
    result,
    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS
  );
}

// Test that on_deactivate returns SUCCESS
TEST_F(RdeLifecycleNodeTest, OnDeactivateReturnsSuccess) {
  auto node = std::make_shared<RdeLifecycleNode>("test_node");
  rclcpp_lifecycle::State state;
  auto result = node->on_deactivate(state);
  EXPECT_EQ(
    result,
    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS
  );
}

// Test that on_cleanup returns SUCCESS
TEST_F(RdeLifecycleNodeTest, OnCleanupReturnsSuccess) {
  auto node = std::make_shared<RdeLifecycleNode>("test_node");
  rclcpp_lifecycle::State state;
  auto result = node->on_cleanup(state);
  EXPECT_EQ(
    result,
    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS
  );
}

// Test that on_shutdown returns SUCCESS
TEST_F(RdeLifecycleNodeTest, OnShutdownReturnsSuccess) {
  auto node = std::make_shared<RdeLifecycleNode>("test_node");
  rclcpp_lifecycle::State state;
  auto result = node->on_shutdown(state);
  EXPECT_EQ(
    result,
    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS
  );
}

// Test that on_error returns SUCCESS
TEST_F(RdeLifecycleNodeTest, OnErrorReturnsSuccess) {
  auto node = std::make_shared<RdeLifecycleNode>("test_node");
  rclcpp_lifecycle::State state;
  auto result = node->on_error(state);
  EXPECT_EQ(
    result,
    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS
  );
}

// Test lifecycle transition: configure -> activate
TEST_F(RdeLifecycleNodeTest, LifecycleConfigureToActivate) {
  auto node = std::make_shared<RdeLifecycleNode>("test_node");
  rclcpp_lifecycle::State state;
  
  // First configure
  auto result = node->on_configure(state);
  EXPECT_EQ(
    result,
    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS
  );
  
  // Then activate
  result = node->on_activate(state);
  EXPECT_EQ(
    result,
    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS
  );
}

// Test lifecycle transition: activate -> deactivate
TEST_F(RdeLifecycleNodeTest, LifecycleActivateToDeactivate) {
  auto node = std::make_shared<RdeLifecycleNode>("test_node");
  rclcpp_lifecycle::State state;
  
  // Activate
  auto result = node->on_activate(state);
  EXPECT_EQ(
    result,
    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS
  );
  
  // Then deactivate
  result = node->on_deactivate(state);
  EXPECT_EQ(
    result,
    rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn::SUCCESS
  );
}

File: samples/lifecycle_py/test/test_lifecycle_node.py
"""
Unit tests for the ROS 2 lifecycle node example (Python).
"""

import unittest
from unittest.mock import patch, MagicMock

import rclpy
from rclpy.lifecycle import State, TransitionCallbackReturn

# Import the lifecycle node class
import sys
import os

# Add the source directory to the path
sys.path.insert(0, os.path.join(os.path.dirname(__file__), '..', 'src'))

from rde_lifecycle_py.rde_lifecycle_py import RdeLifecycleNode


class TestRdeLifecycleNode(unittest.TestCase):
    """Test cases for RdeLifecycleNode."""

    def setUp(self):
        """Set up test fixtures."""
        rclpy.init()

    def tearDown(self):
        """Clean up after tests."""
        rclpy.shutdown()

    def test_node_instantiation(self):
        """Test that the lifecycle node can be instantiated."""
        node = RdeLifecycleNode('test_node')
        self.assertIsNotNone(node)
        self.assertEqual(node.get_name(), 'test_node')

    def test_on_configure_returns_success(self):
        """Test that on_configure returns SUCCESS."""
        node = RdeLifecycleNode('test_node')
        state = State(id=1, label='')
        result = node.on_configure(state)
        self.assertEqual(result, TransitionCallbackReturn.SUCCESS)

    def test_on_activate_returns_success(self):
        """Test that on_activate returns SUCCESS."""
        node = RdeLifecycleNode('test_node')
        state = State(id=1, label='')
        result = node.on_activate(state)
        self.assertEqual(result, TransitionCallbackReturn.SUCCESS)

    def test_on_deactivate_returns_success(self):
        """Test that on_deactivate returns SUCCESS."""
        node = RdeLifecycleNode('test_node')
        state = State(id=1, label='')
        result = node.on_deactivate(state)
        self.assertEqual(result, TransitionCallbackReturn.SUCCESS)

    def test_on_cleanup_returns_success(self):
        """Test that on_cleanup returns SUCCESS."""
        node = RdeLifecycleNode('test_node')
        state = State(id=1, label='')
        result = node.on_cleanup(state)
        self.assertEqual(result, TransitionCallbackReturn.SUCCESS)

    def test_on_shutdown_returns_success(self):
        """Test that on_shutdown returns SUCCESS."""
        node = RdeLifecycleNode('test_node')
        state = State(id=1, label='')
        result = node.on_shutdown(state)
        self.assertEqual(result, TransitionCallbackReturn.SUCCESS)

    def test_on_error_returns_success(self):
        """Test that on_error returns SUCCESS."""
        node = RdeLifecycleNode('test_node')
        state = State(id=1, label='')
        result = node.on_error(state)
        self.assertEqual(result, TransitionCallbackReturn.SUCCESS)

    def test_lifecycle_transitions(self):
        """Test a series of lifecycle transitions."""
        node = RdeLifecycleNode('test_node')
        state = State(id=1, label='')

        # Test configure -> activate -> deactivate -> cleanup sequence
        result = node.on_configure(state)
        self.assertEqual(result, TransitionCallbackReturn.SUCCESS)

        result = node.on_activate(state)
        self.assertEqual(result, TransitionCallbackReturn.SUCCESS)

        result = node.on_deactivate(state)
        self.assertEqual(result, TransitionCallbackReturn.SUCCESS)

        result = node.on_cleanup(state)
        self.assertEqual(result, TransitionCallbackReturn.SUCCESS)

    def test_get_logger(self):
        """Test that the node has a valid logger."""
        node = RdeLifecycleNode('test_node')
        logger = node.get_logger()
        self.assertIsNotNone(logger)


if __name__ == '__main__':
    unittest.main()
